storage/range_tree_test.go

```diff
@@ -167,343 +310,478 @@ func TestRotateRight(t *testing.T) {
 				Key:       keyRight,
-				ParentKey: proto.KeyMin,
-				Black:     false,
+				ParentKey: &keyNode,
+				RightKey:  &keyRightRight,
 			},
 			expectedNode: &proto.RangeTreeNode{
-				Key:       keyRight,
-				ParentKey: proto.KeyMin,
-				Black:     true,
-				LeftKey:   &keyNode,
+				Key:       keyNode,
+				ParentKey: &keyRight,
+				LeftKey:   &keyLeft,
 			},
 			expectedRight: &proto.RangeTreeNode{
-				Key:       keyNode,
-				ParentKey: proto.KeyMin,
-				Black:     false,
+				Key:      keyRight,
+				LeftKey:  &keyNode,
+				RightKey: &keyRightRight,
 			},
-			expectedErr: false,
 		},
 	}
 
 	for i, test := range testCases {
 		tc := createTreeContext(&keyNode, []*proto.RangeTreeNode{
+			test.parent,
 			test.node,
 			test.right,
+			test.rightLeft,
 		})
 
 		// Perform the rotation.
-		actualNode, err := tc.rotateLeft(test.node)
+		actualRight, err := tc.rotateLeft(test.node)
 		if err != nil {
-			// Did we expect the error?
-			if test.expectedErr {
-				continue
-			}
 			t.Fatal(err)
 		}
 
-		// Are both nodes dirty?
-		if !tc.nodes[string(keyNode)].dirty {
-			t.Errorf("%d: Expected node to be dirty", i)
-		}
-		if !tc.nodes[string(keyRight)].dirty {
-			t.Errorf("%d: Expected left node to be dirty", i)
-		}
+		checkTreeNode(t, tc, i, "parent", keyParent, test.expectedParent, nil)
+		checkTreeNode(t, tc, i, "node", keyNode, test.expectedNode, nil)
+		checkTreeNode(t, tc, i, "right", keyRight, test.expectedRight, actualRight)
+		checkTreeNode(t, tc, i, "rightLeft", keyRightLeft, test.expectedRightLeft, nil)
 
-		// Is the Fetched Node the same as the retruend one?
-		actualNodeFetched, err := tc.getNode(&keyRight)
+		// Perform the reverse rotation and expect to get the original values
+		// back.
+		actualNode, err := tc.rotateRight(actualRight)
 		if err != nil {
 			t.Fatal(err)
 		}
-		if !reflect.DeepEqual(actualNodeFetched, actualNode) {
-			t.Errorf("%d: Fetched Node is not the same as the returned one. Fetched: %+v Returned: %+v", i, actualNodeFetched, actualNode)
+
+		checkTreeNode(t, tc, i, "parent-reversed", keyParent, test.parent, nil)
+		checkTreeNode(t, tc, i, "node-reversed", keyNode, test.node, actualNode)
+		checkTreeNode(t, tc, i, "right-reversed", keyRight, test.right, nil)
+		checkTreeNode(t, tc, i, "rightLeft-reversed", keyRightLeft, test.rightLeft, nil)
+	}
+}
+
+// VerifyTree checks to ensure that the tree is indeed balanced and a correct
+// red-black tree. It does so by checking each of the red-black tree properties.
+func VerifyTree(t *testing.T, tc *treeContext, testName string) {
+	root, err := tc.getNode(&tc.tree.RootKey)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Property 1 is always correct. All nodes are already colored.
+	verifyProperty2(t, tc, testName, root)
+	// Property 3 is always correct. All leaves are black.
+	verifyProperty4(t, tc, testName, root)
+	pathBlackCount := new(int)
+	*pathBlackCount = -1
+	verifyProperty5(t, tc, testName, root, 0, pathBlackCount)
```

Comment:
It doesn't look like you actually verify that the tree is a binary tree here, just that the nodes obey red/black rules.